### PR TITLE
optimize code

### PR DESCRIPTION
**Pull request by HydrogenSulfate**
### PR types

Function optimization
### PR changes

APIs 
### Describe

主要修改点:
1. 增加部分 API 的docstring 和 type hints
2. 优化 `metric` 模块下的代码逻辑；统一返回数据类型为 `Tensor`，而非 `float/Tensor` 的混合情况，将浮点数转换移动至 eval 内

其他：
1. 优化部分代码逻辑，减少冗余代码
2. 修复 eval 过程中 batch size 统计不正确的问题，明确 docstring 的描述，删除冗余断言
3. 使用 L-BFGS 优化器时增加对 `update_freq=1` 的检查，更正 `train_LBFGS_epoch_func` 中 `reader_cost` 的正确计时逻辑
4. 取消 `lr_scheduler` 更新与 `update_freq` 的关联
5. 删除无用函数 `_mkdir_if_not_exist`

### CODE REVIEW

File: ppsci/metric/base.py [hunk 1]
```diff
@@ -18,5 +18,6 @@
 class Metric(nn.Layer):
     """Base class for metric."""
 
-    def __init__(self):
+    def __init__(self, keep_batch: str = False):
```
[the remark this answers is not in the document]
Comment (HydrogenSulfate) [reply ppsci/metric/base.py]: Done

**Review by zhiminzhang0830 — APPROVED** [review]
LGTM
